src/waitress/channel.py

```diff
@@ -203,9 +222,12 @@ def received(self, data):
                 break
             data = data[n:]
 
-        if requests:
-            self.requests = requests
-            self.server.add_task(self)
+        with self.task_lock:
+            if requests:
+                self.requests.extend(requests)
+                if not self.task_running:
+                    self.server.add_task(self)
+                    self.task_running = True
```

Comment:
isn't task_running redundant? if there's an item in `self.requests` then a task should be running, and we ensure we don't pop each request until a task is complete

Reply:
This was necessary because `self.requests` might drop to an empty list at the beginning of `service`. But I guess by postponing `self.requests.pop(0)` until later (in the block at the end of `service` inside the `task_lock`), it is not necessary and testing whether `self.requests` is empty or not is sufficient.